wgsl/index.bs

```diff
@@ -955,15 +957,56 @@ An <dfn noexport>array</dfn> is an indexable grouping of element values.
   <thead>
     <tr><th>Type<th>Description
   </thead>
-  <tr><td algorithm="sized array type">array&lt;|E|,|N|&gt;
-      <td>A <dfn>sized array</dfn> with |N| elements of type |E|.<br>
-                |N| must be 1 or larger.
-  <tr><td algorithm="runtime-sized array type">array&lt;|E|&gt;<td>A <dfn noexport>runtime-sized</dfn> array of elements of type |E|.
-                       These may only appear in specific contexts.<br>
+  <tr><td algorithm="fixed-size array type">array&lt;|E|,|N|&gt;
+      <td>A <dfn>fixed-size array</dfn> with |N| elements of type |E|.<br>
+          |N| is called the <dfn>element count</dfn> of the array.
+  <tr><td algorithm="runtime-sized array type">array&lt;|E|&gt;
+      <td>A <dfn noexport>runtime-sized</dfn> array of elements of type |E|.
+          These may only appear in specific contexts.<br>
 </table>
 
-Note: Sized arrays and runtime-sized arrays are always distinct types.
-That is, `array`&lt;|E|,|N|&gt; and `array`&lt;|E|&gt; are different.
+When specified, an element count expression |N| must:
+* be a literal or the name of a [[#module-constants|module-scope constant]], and
+* evaluate to an [=integer scalar=] value greater than zero.
+
+Note: A module-scope constant may be [=pipeline-overridable=].
+
+If the element count expression is a [=pipeline-overridable=] constant,
+then the element count value is fully determined at [=pipeline creation=] time.
+Otherwise the element count value is fully determined at [=shader module creation=] time.
+
+Two array types are the same if and only if all of the following are true:
+* They have the same element type.
+* Their element count specifications match, i.e. any of the following are true:
+    * They are both runtime-sized.
+    * They are both fixed-sized with element counts |N1| and |N2| such that one of the following is true:
```

Comment:
This was a particular design choice which I'm not 100% sold on, so I wanted to highlight it.

The example below shows cases where certain types are forced to be different by these rules.

But this also means that we lose some ability to 'substitute equals for equals'.   
E.g. before:

                 let a : array<i32,4> = array<i32,4>();  / / zero-init

after:

                 let size = 4;
                 let a : array<i32,size> = array<i32,4>();  / /  fails type-check
                 
An alternative formulation would allow size-specifiers to match if they are:
      - Case 1: literals or non-overridable module-scope constants with the same value
      - Case 2: the name of the same overridable constant

We definitely don't want to allow different overridable constants to match, because they can be overridden to different values.   The type check should depend only on the source text.


On the whole, I decided to go with the more restrictive formulation.  We can relax the rules later.  We can't go in the other direction.